sql: add inverted indexes on arrays

jordanlewis opened the pull request:
Closes #43199.

This commit adds inverted index support to arrays. Inverted index
entries are created from arrays by simply encoding a key that contains
the array element's table key encoding. Nulls are not indexed, since in
SQL, ARRAY[1, NULL] @> ARRAY[NULL] returns false.

For example, in a table t(int, int[]) with an inverted index with id 3
on the int[] column the row (10, [1, NULL, 2]) produces 2 index keys:

```
/tableId/3/1/10
/tableId/3/2/10
```

This encoding scheme is much simpler than the one for JSON, since arrays
don't have "paths": their elements are simply ordinary datums.

Release note (sql change): The inverted index implementation now
supports indexing array columns. This permits accelerating containment
queries (@> and <@) on array columns by adding an index to them.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45157)
<!-- Reviewable:end -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45157)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RaduBerinde, and @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 696 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0PZ3N0CLCQ5LCxWXOK:-M0PZ3N0CLCQ5LCxWXOL:b-x0c4uz) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/logictest/testdata/logic_test/inverted_index#L696)):*
> ```
> 
> statement ok
> DROP TABLE table_with_nulls
> ```

Can we also add a few relevant optimizer tests in `opt/xform/testdata/rules/select` in the `GenerateInvertedIndexScans` section? There's overlap with the `execbuilder` tests, but in the optimizer tests we're specifically testing that the logic in `GenerateInvertedIndexScans` handles Array inverted indexes in addition to JSON. So would probably require just a couple sanity check cases at that level.
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 759 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0PWeT6CfDaunwmiBTg:-M0PWeT6CfDaunwmiBTh:bp774vr) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/opt/idxconstraint/index_constraints.go#L759)):*
> ```Go
> }
> 
> func (c *indexConstraintCtx) makeInvertedIndexSpansForJSONExpr(
> ```

Would be good to have at least a simple comment on this method.
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 873 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0PWnCuB8u4a13m6na9:-M0PWnCuB8u4a13m6naA:b-c7qh4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/opt/idxconstraint/index_constraints.go#L873)):*
> ```Go
> 		case types.JsonFamily:
> 			return c.makeInvertedIndexSpansForJSONExpr(nd, constraints, allPaths, rightDatum)
> 		case types.ArrayFamily:
> ```

I'd expect `idxconstraint/testdata` tests for this new functionality. "Onion testing" - test at each layer of abstraction.
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 874 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0PWi28DNKjBP2-TVJl:-M0PWi293zgj_R8hYr26:b-pa6z10) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/opt/idxconstraint/index_constraints.go#L874)):*
> ```Go
> 			return c.makeInvertedIndexSpansForJSONExpr(nd, constraints, allPaths, rightDatum)
> 		case types.ArrayFamily:
> 			// We're going to make one span to search for every value inside of the
> ```

I think this code should be broken out into a helper method, analogous to the JSON method.
___
*[pkg/sql/sqlbase/index_encoding.go, line 1237 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0PXirBAMe0tZbIg7F1:-M0PXirBAMe0tZbIg7F2:b-ug1eg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/sqlbase/index_encoding.go#L1237)):*
> ```Go
> 			// have 0 or >1 entries, though, as well as secondary indexes which store
> 			// columns from multiple column families.
> 			secondaryIndexEntries = append(secondaryIndexEntries, entries[i])
> ```

Wouldn't this just be:
```
secondaryIndexEntries = append(secondaryIndexEntries, entries)
```


<!-- Sent from Reviewable.io -->

rohany reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45157)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RaduBerinde, and @rohany)

___
*[pkg/cli/testdata/dump/inverted_index, line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0OeOXJ3LdrXYjTdT2l:-M0OeOXJ3LdrXYjTdT2m:bq5tnk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/85648a79113a812bda4dc79e4a53a905a9dbdc72/pkg/cli/testdata/dump/inverted_index#L6)):*
> ```
> 	a JSON,
> 	b JSON,
>     c INT[],
> ```

the spacing seems off here (missing tabs or something)
___
*[pkg/cli/testdata/dump/inverted_index, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0Ojcbr5Zh4LDxn-9LO:-M0Ojcbr5Zh4LDxn-9LP:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/85648a79113a812bda4dc79e4a53a905a9dbdc72/pkg/cli/testdata/dump/inverted_index#L25)):*
> ```
> 	a JSONB NULL,
> 	b JSONB NULL,
>     c INT[] NULL,
> ```

same as above
___
*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 716 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QQj964hcahL92R7Q6:-M0QQj964hcahL92R7Q7:b-t0lnjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/logictest/testdata/logic_test/inverted_index#L716)):*
> ```
> 
> statement ok
> CREATE INDEX ON c USING GIN (bar)
> ```

add a test where the backfiller runs and has to add non null array rows
___
*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 749 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QPeX91VHsq6ZNJciC:-M0QPeX91VHsq6ZNJciD:bse30rx) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/logictest/testdata/logic_test/inverted_index#L749)):*
> ```
> 
> query error unsupported comparison operator
> SELECT * FROM c WHERE foo @> 0
> ```

I'm confused about this syntax. So we can only test containment for singleton arrays? or would be in the future allow for containment checks of non singleton arrays? I find it unintuitive that we have to do `array @> array[element]` rather than `array @> element`.
___
*[pkg/sql/row/updater.go, line 382 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QQM_gCaO8FUGm5wZo:-M0QQM_gCaO8FUGm5wZp:b4ntl5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/row/updater.go#L382)):*
> ```Go
> 			ru.Fks.addCheckForIndex(ru.Helper.TableDesc.PrimaryIndex.ID, ru.Helper.TableDesc.PrimaryIndex.Type)
> 			for i := range ru.Helper.Indexes {
> 				if ru.Helper.Indexes[i].Type == sqlbase.IndexDescriptor_INVERTED {
> ```

Was this missing / or deleted accidentally?
___
*[pkg/sql/sem/builtins/builtins.go, line 3634 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QQsfb1-e9et_38RKn:-M0QQsfb1-e9et_38RKo:b-r0qmok) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/sem/builtins/builtins.go#L3634)):*
> ```Go
> 		},
> 		tree.Overload{
> 			Types:      tree.ArgTypes{{"val", types.AnyArray}},
> ```

Is it important to test this function explicitly? Also, might be worth to drop a comment in the key encoder to update this if the implementation changes.
___
*[pkg/sql/sqlbase/index_encoding.go, line 879 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QR1mA0UN7Pt4huUmQ:-M0QR1mA0UN7Pt4huUmR:b-rn5c2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/sqlbase/index_encoding.go#L879)):*
> ```Go
> // prefixed to all returned keys.
> // N.B.: This won't return any keys for
> func encodeArrayInvertedIndexTableKeys(val *tree.DArray, inKey []byte) (key [][]byte, err error) {
> ```

I'm making a change to `EncodeTableKey` in https://github.com/cockroachdb/cockroach/pull/44771. I don't think it should impact this work, but it's worth a double check.
___
*[pkg/sql/sqlbase/index_encoding.go, line 893 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QRMFP3QxOWlvEbIE2:-M0QRMFP3QxOWlvEbIE3:bs4a99e) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/sqlbase/index_encoding.go#L893)):*
> ```Go
> 			return nil, err
> 		}
> 		outKey := make([]byte, len(inKey), len(inKey)+len(newKey))
> ```

Can we allocate the new outkey, copy inKey into it, and then encodeTableKey into that buffer? Right now we are allocating outkey and newKey for no reason.
___
*[pkg/sql/sqlbase/index_encoding.go, line 1235 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QRlAD4x_PXgYynw7-:-M0QRlAD4x_PXgYynw70:bj0h8wt) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/sqlbase/index_encoding.go#L1235)):*
> ```Go
> 		for i := range entries {
> 			// Normally, each index will have exactly one entry. Inverted indexes can
> 			// have 0 or >1 entries, though, as well as secondary indexes which store
> ```

This sentence seems wonky


<!-- Sent from Reviewable.io -->

jordanlewis committed:
sql: add inverted indexes on arrays

This commit adds inverted index support to arrays. Inverted index
entries are created from arrays by simply encoding a key that contains
the array element's table key encoding. Nulls are not indexed, since in
SQL, ARRAY[1, NULL] @> ARRAY[NULL] returns false.

For example, in a table t(int, int[]) with an inverted index with id 3
on the int[] column the row (10, [1, NULL, 2]) produces 2 index keys:

```
/tableId/3/1/10
/tableId/3/2/10
```

This encoding scheme is much simpler than the one for JSON, since arrays
don't have "paths": their elements are simply ordinary datums.

Release note (sql change): The inverted index implementation now
supports indexing array columns. This permits accelerating containment
queries (@> and <@) on array columns by adding an index to them.

jordanlewis reviewed (commented):
Thanks for the reviews, please take another look. I split a bunch of this PR into #45564, so that one needs to be reviewed first.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-:-M1Nw_gP4erm8Q8YBt1G:bh7hpaz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jordanlewis, @RaduBerinde, and @rohany)

___
*[pkg/cli/testdata/dump/inverted_index, line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0OeOXJ3LdrXYjTdT2l:-M0QdYje947RorLuzOee:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/85648a79113a812bda4dc79e4a53a905a9dbdc72/pkg/cli/testdata/dump/inverted_index#L6)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

the spacing seems off here (missing tabs or something)
</blockquote></details>

Done.
___
*[pkg/cli/testdata/dump/inverted_index, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0Ojcbr5Zh4LDxn-9LO:-M0QdYWw1aKRHvV-GCip:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/85648a79113a812bda4dc79e4a53a905a9dbdc72/pkg/cli/testdata/dump/inverted_index#L25)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 696 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0PZ3N0CLCQ5LCxWXOK:-M1NN_1k54tTXFe1jRsY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/logictest/testdata/logic_test/inverted_index#L696)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Can we also add a few relevant optimizer tests in `opt/xform/testdata/rules/select` in the `GenerateInvertedIndexScans` section? There's overlap with the `execbuilder` tests, but in the optimizer tests we're specifically testing that the logic in `GenerateInvertedIndexScans` handles Array inverted indexes in addition to JSON. So would probably require just a couple sanity check cases at that level.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 716 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QQj964hcahL92R7Q6:-M0QdXACD-139Og_OGMw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/logictest/testdata/logic_test/inverted_index#L716)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

add a test where the backfiller runs and has to add non null array rows
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 749 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QPeX91VHsq6ZNJciC:-M0Qd_IZ8qjPov40nke6:b-wcz2vy) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/logictest/testdata/logic_test/inverted_index#L749)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I'm confused about this syntax. So we can only test containment for singleton arrays? or would be in the future allow for containment checks of non singleton arrays? I find it unintuitive that we have to do `array @> array[element]` rather than `array @> element`.
</blockquote></details>

Yeah, this is just how `@>` works - it's funky. Contains in Postgres world compares the container type to itself. array @> element isn't a supported comparison.

array @> array is also supported for non singletons on the RHS, which means that the LHS must contain all elements in the RHS. Under the hood we turn that into a conjunction of singleton @> (i think) - and it turns into a zigzag join. I'll add some tests for that case as well.
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 759 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0PWeT6CfDaunwmiBTg:-M0TY0iKACOr4SxLWc1o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/opt/idxconstraint/index_constraints.go#L759)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Would be good to have at least a simple comment on this method.
</blockquote></details>

Done.
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 874 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0PWi28DNKjBP2-TVJl:-M0TY1nSELzrQ4nuYkZX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/opt/idxconstraint/index_constraints.go#L874)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think this code should be broken out into a helper method, analogous to the JSON method.
</blockquote></details>

Done.
___
*[pkg/sql/row/updater.go, line 382 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QQM_gCaO8FUGm5wZo:-M1N6y1t2W3G-0QJCn6g:b-ws8qaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/row/updater.go#L382)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Was this missing / or deleted accidentally?
</blockquote></details>

I think it was already implicitly true via another check that I got rid of.
___
*[pkg/sql/sem/builtins/builtins.go, line 3634 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QQsfb1-e9et_38RKn:-M0TZKoNA6WoPey9HurU:bp5xtdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/sem/builtins/builtins.go#L3634)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Is it important to test this function explicitly? Also, might be worth to drop a comment in the key encoder to update this if the implementation changes.
</blockquote></details>

Good idea. This actually caught a bug - thanks for pointing it out.
___
*[pkg/sql/sqlbase/index_encoding.go, line 879 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QR1mA0UN7Pt4huUmQ:-M1N76yIDAPNlr_5bdZb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/sqlbase/index_encoding.go#L879)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I'm making a change to `EncodeTableKey` in https://github.com/cockroachdb/cockroach/pull/44771. I don't think it should impact this work, but it's worth a double check.
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/index_encoding.go, line 893 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QRMFP3QxOWlvEbIE2:-M1NAx0l3aaSuqYqeamy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/sqlbase/index_encoding.go#L893)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Can we allocate the new outkey, copy inKey into it, and then encodeTableKey into that buffer? Right now we are allocating outkey and newKey for no reason.
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/index_encoding.go, line 1235 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0QRlAD4x_PXgYynw7-:-M1N7U832_YXYUuj6Fu9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/sqlbase/index_encoding.go#L1235)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

This sentence seems wonky
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/index_encoding.go, line 1237 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0PXirBAMe0tZbIg7F1:-M1N7LfH5emVut433ze8:bhs3g3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/sqlbase/index_encoding.go#L1237)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Wouldn't this just be:
```
secondaryIndexEntries = append(secondaryIndexEntries, entries)
```
</blockquote></details>

I think you need `...` syntax, but yes, that's a good point - fixed.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45157)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @rohany)

___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 873 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0PWnCuB8u4a13m6na9:-M1OZE4Q1VaCIrbs4sgs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/afc64560172986ffc6bb0d5c12e301371a9d4d08/pkg/sql/opt/idxconstraint/index_constraints.go#L873)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'd expect `idxconstraint/testdata` tests for this new functionality. "Onion testing" - test at each layer of abstraction.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rohany commented on pkg/sql/opt/exec/execbuilder/testdata/inverted_index:
```diff
@@ -78,6 +96,52 @@ DELETE FROM d WHERE a=4
+
+# Make sure that empty arrays do not emit any keys at all.
+query T kvtrace
+INSERT INTO e VALUES(1, ARRAY[])
```
what happens with array of null -- do we emit a key there?

rohany commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -3622,6 +3623,33 @@ may increase either contention or retry errors, or both.`,
 			},
 			Info: "This function is used only by CockroachDB's developers for testing purposes.",
 		},
+		tree.Overload{
```
This seems like it's out of date given we don't emit keys for duplicate entries

rohany reviewed (changes requested):
This needs a version gate as well

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45157)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @rohany)

___
*[pkg/sql/opt/exec/execbuilder/testdata/inverted_index, line 113 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0Le9uL9ikRBf9LM5Uf-r6-113:-M1Z0-o1BGH93j4mi7K6:bduebru) ([raw file](https://github.com/cockroachdb/cockroach/blob/dae039f94f4a54abf57cfea0d247ad808d6738ff/pkg/sql/opt/exec/execbuilder/testdata/inverted_index#L113)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

what happens with array of null -- do we emit a key there?
</blockquote></details>

No, we don't. Added tests. It also caught another bug, so thanks for asking!
___
*[pkg/sql/sem/builtins/builtins.go, line 3626 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0Le9uMrFrNCifEabg9-r6-3626:-M1Z049oBAv_FEl0ualr:b1diu57) ([raw file](https://github.com/cockroachdb/cockroach/blob/dae039f94f4a54abf57cfea0d247ad808d6738ff/pkg/sql/sem/builtins/builtins.go#L3626)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

This seems like it's out of date given we don't emit keys for duplicate entries
</blockquote></details>

What do you mean? I think it's still correct.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
I don't think it needs a version gate. Schema changes won't be permitted in the mixed version state. Unless there's something else you're thinking of?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-:-M1Z3aIa4Vh-08EMo2Yi:b-vte740)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @rohany)



<!-- Sent from Reviewable.io -->

rohany reviewed (changes requested):
Yeah, assuming we don't allow schema changes in a mixed version state then its ok. I was just worried about the 20.1 nodes in a mixed version setting creating an inverted index on an array column.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-:-M1a3NjdC07UkHUuB5gF:b-woj6tq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jordanlewis, @RaduBerinde, and @rohany)

___
*[pkg/sql/sem/builtins/builtins.go, line 3626 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0Le9uMrFrNCifEabg9-r6-3626:-M1a3H5o4DN9H4mfRTvA:b7eax6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/dae039f94f4a54abf57cfea0d247ad808d6738ff/pkg/sql/sem/builtins/builtins.go#L3626)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What do you mean? I think it's still correct.
</blockquote></details>

The count here is counting all elements that aren't null, while we should be counting all unique elements right?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45157)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @rohany)

___
*[pkg/sql/sem/builtins/builtins.go, line 3626 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0Le9uMrFrNCifEabg9-r6-3626:-M1a4R5xB-gHYbpf1E28:bukh706) ([raw file](https://github.com/cockroachdb/cockroach/blob/dae039f94f4a54abf57cfea0d247ad808d6738ff/pkg/sql/sem/builtins/builtins.go#L3626)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

The count here is counting all elements that aren't null, while we should be counting all unique elements right?
</blockquote></details>

You're right...  how did I miss this? Added tests.


<!-- Sent from Reviewable.io -->

rohany commented:
SQL parts LGTM -- lets wait for an LGTM from someone on the optimizer team before merging

jordanlewis reviewed (commented):
@andy-kimball, would you have a moment to take another look at this? Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-:-M1fRRfJALRvuKfVYg8Q:b-tvlpjf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @rohany)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Opt part LGTM (with some comments)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-:-M1fcXo97NadO6MAIphP:b-9j1ytw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jordanlewis, @RaduBerinde, and @rohany)

___
*[pkg/cli/testdata/dump/inverted_index, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0OgQA323zxiPCPw-si:-M0OgQA488Va8E18al2q:bg0bw7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/85648a79113a812bda4dc79e4a53a905a9dbdc72/pkg/cli/testdata/dump/inverted_index#L24)):*
> ```
> CREATE TABLE t (
> 	a JSONB NULL,
> 	b JSONB NULL,
> ```

[nit] fix the tabs in the rest of the query
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 845 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1fbsud-gM-A6UliMMX:-M1fbsud-gM-A6UliMMY:beokzmj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/idxconstraint/index_constraints.go#L845)):*
> ```Go
> // makeInvertedIndexSpans for array inverted indexes. The input arr is the array
> // to produce spans for.
> func (c *indexConstraintCtx) makeInvertedIndexSpansForArrayExpr(
> ```

[nit] include a description of `allPaths`
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 853 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1fb128DjkS-oHK2t5T:-M1fb128DjkS-oHK2t5U:b-sios6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/idxconstraint/index_constraints.go#L853)):*
> ```Go
> 	// array datum.
> 	for i := range arr.Array {
> 		elt := arr.Array[i]
> ```

[nit] `elt` is used once, I'd just use `arr.Array[i]`
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 856 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1fZYwLDmYLUzT7LGuM:-M1fZYwLDmYLUzT7LGuN:bx8w3p3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/idxconstraint/index_constraints.go#L856)):*
> ```Go
> 		if elt == tree.DNull {
> 			// In SQL:
> 			// SELECT ARRAY[1, NULL, 2] @> ARRAY[NULL]
> ```

[nit] this example is confusing, it's not clear which side our NULL corresponds to. Maybe just say that in SQL, testing an array that contains NULL for containment in another array always yields false.
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 871 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1fbG2kC36ughYMWDgG:-M1fbG2kC36ughYMWDgH:b-vpx8fr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/idxconstraint/index_constraints.go#L871)):*
> ```Go
> 		if !allPaths {
> 			// The span is tight if we just had 1 path through the index constraint.
> 			return len(arr.Array) == 1, constraints
> ```

[nit] This can just be `break` and we would have a single instance of the tightness logic.
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 875 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1faEUD4lSizLKvj2wm:-M1faEUD4lSizLKvj2wn:b-8ugnhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/idxconstraint/index_constraints.go#L875)):*
> ```Go
> 
> 		// Reset out for next iteration
> 		out = &constraint.Constraint{}
> ```

We can just do `out := &constraint.Constraint()` above where we use it. I know this is copied but there we had a `continue` case.
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 878 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1f_yYPBWwHwjMh1YKc:-M1f_yYPBWwHwjMh1YKd:bhrv8wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/idxconstraint/index_constraints.go#L878)):*
> ```Go
> 	}
> 	if !constrained {
> 		c.unconstrained(0 /* offset */, out)
> ```

[nit] This is only possible if it's an empty array right? We could just check for that upfront
___
*[pkg/sql/opt/idxconstraint/testdata/inverted, line 78 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1fcQQt2-Mz4uWOboxa:-M1fcQQt2-Mz4uWOboxb:bnfflbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/idxconstraint/testdata/inverted#L78)):*
> ```
> [ - ]
> Remaining filter: @1 @> ARRAY[]
> ```

[nit] add some tests for the case where the array contains a NULL
___
*[pkg/sql/opt/idxconstraint/testdata/inverted, line 83 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1fbjsMBZ8bDQjjBlr9:-M1fbjsMBZ8bDQjjBlrA:b-ijjemz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/idxconstraint/testdata/inverted#L83)):*
> ```
> ----
> [ - ]
> Remaining filter: NULL
> ```

This should be a contradiction
___
*[pkg/sql/opt/xform/testdata/rules/select, line 932 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1fbzO29nqQEmoATN1R:-M1fbzO29nqQEmoATN1S:b-ljqrbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/xform/testdata/rules/select#L932)):*
> ```
>            └── key: (1)
> 
> # Write some tests for array inverted indexes.
> ```

[nit] remove "write some"


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45157)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jordanlewis, @RaduBerinde, and @rohany)

___
*[pkg/cli/testdata/dump/inverted_index, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0OgQA323zxiPCPw-si:-M1gB0jUBeA9SjiPf3SZ:b-9tk1fz) ([raw file](https://github.com/cockroachdb/cockroach/blob/85648a79113a812bda4dc79e4a53a905a9dbdc72/pkg/cli/testdata/dump/inverted_index#L24)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] fix the tabs in the rest of the query
</blockquote></details>

I think there's nothing wrong with them anymore. What am i missing?
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 875 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1faEUD4lSizLKvj2wm:-M1gC49JEQy58jk-WtbI:b-l97i80) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/idxconstraint/index_constraints.go#L875)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We can just do `out := &constraint.Constraint()` above where we use it. I know this is copied but there we had a `continue` case.
</blockquote></details>

Good point, fixed.
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 878 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1f_yYPBWwHwjMh1YKc:-M1gDBwB-4JKif_OOZ2M:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/idxconstraint/index_constraints.go#L878)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] This is only possible if it's an empty array right? We could just check for that upfront
</blockquote></details>

Good point, done.
___
*[pkg/sql/opt/idxconstraint/testdata/inverted, line 83 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1fbjsMBZ8bDQjjBlr9:-M1gDGO_0LJbUPFsWz2b:bgmuu5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/idxconstraint/testdata/inverted#L83)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This should be a contradiction
</blockquote></details>

Yeah true. I added tests for this for JSON too and it's also broken. I think there's something about the way that this test is set up that makes this not work properly.
___
*[pkg/sql/opt/xform/testdata/rules/select, line 932 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M1fbzO29nqQEmoATN1R:-M1gKYjj60HxsWOeNIUU:bo489gt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5b68740c7628859281a6629e0ddd904982cd261/pkg/sql/opt/xform/testdata/rules/select#L932)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] remove "write some"
</blockquote></details>

Done. this was a weird comment


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-:-M1gOmA6D7AywGerEo6C:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @jordanlewis, @RaduBerinde, and @rohany)

___
*[pkg/cli/testdata/dump/inverted_index, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45157#-M0OgQA323zxiPCPw-si:-M1gO-iW2n5VV8ssNTnh:b-xgvk0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/85648a79113a812bda4dc79e4a53a905a9dbdc72/pkg/cli/testdata/dump/inverted_index#L24)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think there's nothing wrong with them anymore. What am i missing?
</blockquote></details>

Hm I don't remember even writing this haha


<!-- Sent from Reviewable.io -->

bra-fsn commented:
I guess this closes https://github.com/cockroachdb/cockroach/issues/17154 as well.
